### PR TITLE
Cumulus 3357: Convert queue-granules to typescript

### PR DESCRIPTION
**Pull request by sbquinlan**
**Summary:** Summary of changes

Addresses [CUMULUS-3357](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3357)

## Changes

* Added an option to `@culumus/schemas` to generate typescript types from the json schema files for a task.
* Converted the queue-granules task to use typescript

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: tasks/queue-granules/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
+/**
+ * This file is generated using @cumulus/schema. Any modifications made to this file
+ * will the overwritten when the build script is rerun. Please do not modify this file.
```
Comment (botanical) [inline tasks/queue-granules/src/types.ts:3]:
```suggestion
 * will be overwritten when the build script is rerun. Please do not modify this file.
```

File: tasks/queue-granules/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    ...granule,
+    createdAt: granule.createdAt ? granule.createdAt : createdAt,
+  }));
+}
+
+/**
+ * See schemas/input.json and schemas/config.json for detailed event description
+ *
+ * @param {HandlerEvent} event - Lambda event object
+ * @param {Object} testMocks - Object containing mock functions for testing
```
Comment (botanical) [inline tasks/queue-granules/src/index.ts:101]:
```suggestion
```

File: tasks/queue-granules/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+export interface QueueGranulesInput {
+  pdr?: {
+    name: string;
+    path: string;
+  };
+  granules: {
+    type?: string;
+    granuleId: string;
+    dataType?: string;
+    version?: string;
```
Comment (botanical) [inline tasks/queue-granules/src/types.ts:35]: Are dataType and version not required?

File: tasks/queue-granules/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  internalBucket: string;
+  provider: {
+    id: string;
+  };
+  preferredQueueBatchSize?: number | null;
+  queueUrl: string;
+  stackName: string;
+  concurrency?: number;
+  executionNamePrefix?: string;
+  childWorkflowMeta?: {};
```
Comment (botanical) [inline tasks/queue-granules/src/types.ts:20]: what's the difference between `childWorkflowMeta?: {}` and `childWorkflowMeta?: object`?

File: CHANGELOG.md [hunk 5]
```diff
@@ -29,6 +29,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     granules at a time. 
   - Switched `@cumulus/queue-granules` to use the `bulkUpdateGranules` api instead of individually
     updating one granule at a time to increase throughput to the database.
+- **CUMULUS-3357**
+  - `@cumulus/queue-granules` is now written in Typescript
```
Comment (botanical) [inline CHANGELOG.md:33]:
```suggestion
  - `@cumulus/queue-granules` is now written in TypeScript
```
a nit

File: CHANGELOG.md [hunk 6]
```diff
@@ -29,6 +29,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
+  - `@cumulus/schemas` can now generate Typescript interfaces for the task input, output and config.
```
Comment (botanical) [inline CHANGELOG.md:34]:
```suggestion
  - `@cumulus/schemas` can now generate TypeScript interfaces for the task input, output and config.
```

File: packages/schemas/src/generate-types.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import fs from 'fs/promises';
+import path from 'path';
+
+import type { FileHandle } from 'fs/promises';
+import type { JSONSchema } from 'json-schema-to-typescript';
+
+const BANNER_COMMENT = `
+/**
+ * This file is generated using @cumulus/schema. Any modifications made to this file
+ * will the overwritten when the build script is rerun. Please do not modify this file.
```
Comment (botanical) [inline packages/schemas/src/generate-types.ts:13]:
```suggestion
 * will be overwritten when the build script is rerun. Please do not modify this file.
```

File: tasks/queue-granules/src/types.ts (hunk 3 above)
Comment (botanical) [reply tasks/queue-granules/src/types.ts:35]: I don't think they are 😅 so disregard

**Review by botanical — APPROVED** [review]
Had a couple nit comments. The conversion to TS looks solid ⭐

File: tasks/queue-granules/src/types.ts (hunk 4 above)
Comment (sbquinlan) [reply tasks/queue-granules/src/types.ts:20]: https://stackoverflow.com/questions/49464634/difference-between-object-and-object-in-typescript

Seems like `object` is probably best though not that different from `{}`, but `object` is probably a better fit. I don't think I can make that change without writing my own version of json-schema-to-typescript

**Pull request closed without merging**